[Feature] Support Hive csv array.

Smith-Cruise opened the pull request:
## What type of PR is this：
- [x] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #7780

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->
Support Hive array format, and be compatible with the original array converter.

Smith-Cruise committed:
[Feature] Support hive csv array.

Smith-Cruise committed:
[Feature] Support hive csv array.

DorianZheng commented on be/src/formats/csv/converter.cpp:
```diff
@@ -75,4 +75,57 @@ std::unique_ptr<Converter> get_converter(const TypeDescriptor& type_desc, bool n
+        tmp = static_cast<int32_t>(nested_array_level) + (-128 - 25);
+    }
+
+    return (char)tmp;
```
use static_cast

blackstar-baba committed:
[Bugfix] Fix mv version map issues (#8799)

Fix the problem that the materialized view cannot add partitions due to base table replace partition
Fix the problem that the version map was written to an incorrect version (wait for the insert overwrite to be completed and then write the version, and add a lock to ensure the correctness of the write)
Fix partition name relationships add and remove issues

hiliuxg committed:
[SR SQL Planner Summer Camp][Feature] alter table statement support of add list partition (#9019)

imay commented on be/src/exec/vectorized/hdfs_scanner_text.cpp:
```diff
@@ -99,17 +99,29 @@ Status HdfsScannerCSVReader::_fill_buffer() {
+    _mapkey_delimiter = text_file_desc.mapkey_delim.front();
+
+    // _collection_delimiter and _mapkey_delimiter won't be empty
+    CHECK(!_collection_delimiter.empty());
```
Must use DCHECK, we can't let process crash because of one request.

DorianZheng commented on be/src/exec/vectorized/hdfs_scanner_text.cpp:
```diff
@@ -99,17 +99,29 @@ Status HdfsScannerCSVReader::_fill_buffer() {
+
+    // we should cast string to char now since csv reader only support record, collection, mapkey delimiter by char
     _record_delimiter = text_file_desc.line_delim.front();
+    _collection_delimiter = text_file_desc.collection_delim.front();
```
Is delimiter only allowed to be one char?

Smith-Cruise replied:
Yes, I have already checked it.
You can refer to: https://github.com/apache/hive/blob/master/serde/src/java/org/apache/hadoop/hive/serde2/lazy/LazySerDeParameters.java#L251
The separator only occupied one byte.

DorianZheng commented on be/src/exec/vectorized/hdfs_scanner_text.cpp:
```diff
@@ -99,17 +99,29 @@ Status HdfsScannerCSVReader::_fill_buffer() {
-    for (int i = 0; i < _scanner_params.materialize_slots.size(); i++) {
-        auto slot = _scanner_params.materialize_slots[i];
-        ConverterPtr conv = csv::get_converter(slot->type(), true);
+    for (auto slot : _scanner_params.materialize_slots) {
```
const auto&

stdpain committed:
[Bugfix] fix "unsupport decode_dict_codes" error in late_materized (#9046)

After PR #8869, GlobalDictCodeColumnIterator won't support `decode_dict_codes(const int32_t* codes, size_t size, vectorized::Column* words)`,
we need to decode_dict_codes(const vectorized::Column& codes, vectorized::Column* words) instead it

DorianZheng commented on be/src/formats/csv/converter.h:
```diff
@@ -74,6 +74,8 @@ class Converter {
 };
 
 std::unique_ptr<Converter> get_converter(const TypeDescriptor& type_desc, bool nullable);
+std::unique_ptr<Converter> get_converter(const TypeDescriptor& type_desc, bool nullable, char collection_delimiter,
```
I think one `get_converter` is enough, you can give `collection_delimiter` a default value.

caneGuy replied:
+1

> I think one `get_converter` is enough, you can give `collection_delimiter` a default value.

Smith-Cruise replied:
I had to rewrite a new version. PTAL.

stdpain committed:
[Bugfix] Fix DictMappingExpr rewrite multi times (#8950)

Because Expr and ExprContext are shared in OlapScanOperator.
And OlapScanPrepareOperator::pull_chunk() will be executed concurrently,
which can cause the rewrite function to be executed multiple times.
We should use std::call_once to avoid multiple rewrites

dirtysalt commented on be/src/exec/vectorized/hdfs_scanner_text.cpp:
```diff
@@ -99,17 +99,29 @@ Status HdfsScannerCSVReader::_fill_buffer() {
-        ConverterPtr conv = csv::get_converter(slot->type(), true);
+    for (auto slot : _scanner_params.materialize_slots) {
+        ConverterPtr conv;
+        if (slot->type().type == TYPE_ARRAY) {
```
check nullptr here. Possibly there is nullptr.

Smith-Cruise replied:
Do you mean to check that `slot` is empty?

dirtysalt replied:
For some non-existed column, slot will be nullptr. You'd better to skip it.

caneGuy replied:
may be materialize_slots can not be nullptr? @dirtysalt Please correct me if I understand wrong

dirtysalt replied:
slot could be nullptr. I'm not 100% sure root cause of it, but I've come across this case before.

dirtysalt commented on be/src/formats/csv/array_reader.cpp:
```diff
@@ -0,0 +1,109 @@
+}
+
+bool HiveTextArrayReader::read_string(const std::unique_ptr<Converter>& elem_converter, Column* column, Slice s, const Converter::Options& options) const {
+    auto* array = down_cast<ArrayColumn*>(column);
```
almost same logic to StandardArrayrReader?

dirtysalt replied:
maybe this piece of code is common:
1. split array elements into vector of slices
2. use elemnt_converted to parse slice

Smith-Cruise replied:
The array format of Hive is very different from the standard array format. So writing `read_string` separately here will be easier to maintain in the future.

Smith-Cruise replied:
Did you notice that in `StandardArrayReader` is use `read_quoted_string()`, and in `HiveTextArrayReader` is use `read_string`.

dirtysalt replied:
OK. For one reader,  is only one of them `read_string` or `read_quoted_string` be used? If only one is used, possbily we could unify it ?

Smith-Cruise replied:
I had to rewrite a new version. PTAL.

dirtysalt commented on be/src/formats/csv/converter.cpp:
```diff
@@ -75,4 +75,57 @@ std::unique_ptr<Converter> get_converter(const TypeDescriptor& type_desc, bool n
+        tmp = static_cast<int32_t>(nested_array_level) + (-128 - 25);
+    }
+
+    return (char)tmp;
```
static_cast

Smith-Cruise replied:
[DorianZheng](https://github.com/DorianZheng) already mentioned it. 🤣

dirtysalt commented on be/src/exec/vectorized/hdfs_scanner_text.cpp:
```diff
@@ -99,17 +99,29 @@ Status HdfsScannerCSVReader::_fill_buffer() {
+
+    // we should cast string to char now since csv reader only support record, collection, mapkey delimiter by char
     _record_delimiter = text_file_desc.line_delim.front();
+    _collection_delimiter = text_file_desc.collection_delim.front();
```
I guess w'd better to change it to `char`, no more implicit meaning.  delimiter naturally should be character not string.

@caneGuy @DorianZheng @sduzh  what do you think?

Smith-Cruise replied:
The delimiter is an optional field in Thrift so it may be empty.

dirtysalt replied:
Oh, for that reson, we have to make a character: so we don't be confused about state, and use a explicit character to mark it as INVAID_STATE.

caneGuy replied:
actually hive has defined these delimiter as string, So I followed this implementation. https://github.com/apache/hive/blob/90428cc5f594bd0abb457e4e5c391007b2ad1cb8/serde/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/serde/serdeConstants.java#L34-L40

dirtysalt replied:
these items are keys.
<img width="547" alt="image" src="https://user-images.githubusercontent.com/1081215/180587627-43c541bf-0433-48d4-bac3-83f0a101f1cd.png">

dulong41 committed:
Update add-approve-label.yml (#9064)

add workflow run_id to get the most recent pr information

Smith-Cruise committed:
Fix

Smith-Cruise committed:
Fix

blackstar-baba committed:
mv map persistence (#8879)

DorianZheng commented on be/src/formats/csv/converter.cpp:
```diff
@@ -75,4 +75,60 @@ std::unique_ptr<Converter> get_converter(const TypeDescriptor& type_desc, bool n
 
+// Hive collection delimiter generate rule refer to:
+// https://github.com/apache/hive/blob/master/serde/src/java/org/apache/hadoop/hive/serde2/lazy/LazySerDeParameters.java#L250
+static char get_collection_delimiter(char mapkey_delimiter, size_t nested_array_level) {
```
should add ut for this function individually, every util function should have their own ut

DorianZheng commented on be/src/formats/csv/converter.cpp:
```diff
@@ -75,4 +75,60 @@ std::unique_ptr<Converter> get_converter(const TypeDescriptor& type_desc, bool n
+
+    // tmp maybe negative, dont use size_t.
+    int32_t tmp;
+    if (nested_array_level <= 7) {
```
plz add comments

Smith-Cruise replied:
I add detailed comments in `array_converter.cpp`.

DorianZheng commented on fe/fe-core/src/main/java/com/starrocks/external/hive/HiveMetaClient.java:
```diff
@@ -649,10 +649,14 @@ private double getLengthFromLiteral(LiteralExpr literalExpr, Type type) {
-        // get properties like 'field.delim' and 'line.delim' from StorageDescriptor
+        // Get properties 'field.delim', 'line.delim', 'collection.delim' and 'mapkey.delim' from StorageDescriptor
+        // Detail refer to:
+        // https://github.com/apache/hive/blob/master/serde/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/serde/serdeConstants.java#L34-L40
```
suggest use permanent link of github codes

DorianZheng commented on be/src/formats/csv/converter.cpp:
```diff
@@ -75,4 +75,60 @@ std::unique_ptr<Converter> get_converter(const TypeDescriptor& type_desc, bool n
 }
 
+// Hive collection delimiter generate rule refer to:
+// https://github.com/apache/hive/blob/master/serde/src/java/org/apache/hadoop/hive/serde2/lazy/LazySerDeParameters.java#L250
```
suggest use permanent link of github codes

DorianZheng commented on gensrc/thrift/Descriptors.thrift:
```diff
@@ -64,6 +64,12 @@ struct TTextFileDesc {
+    3: optional string collection_delim
+
+    // property 'mapkey.delim'
+    4: optional string mapkey_delim
```
comments what is `mapkey`

Smith-Cruise replied:
I explain it in `converter.h`.

liuyehcf committed:
[Enhancement] Add invisible session variable 'profile_timeout' (#8999)

* [Enhancement] increase profile's waiting timeout

* set timeout to 1 minute

* update comments

* add invisiable session variable

* naming

HangyuanLiu committed:
[Enhancement] Optimized statistics scheduling and not collecting empty tables and partitions (#8958)

Smith-Cruise committed:
Fix unsable Slack links for new users. (#9067)

If you are a new Slack user, the original URL is invalid.

Smith-Cruise committed:
Fix, and add ut for converter.

Smith-Cruise committed:
Fix, and add ut for converter.

HangyuanLiu committed:
[Enhancement] Refactored statistics expired logic and support Drop statistics statement (#8960)

choury committed:
Skip create agent server when start cn (#9061)

Pslydhh committed:
Add compress debug info when disable minidump (#9043)

Compress debug info to reduce file size(1.7GB) to 600MB.

xlwh committed:
[Refactor] Remove old parser for partition & function related statements (#9083)

dulong41 committed:
[Actions] fix pr and run_id cannot use together (#9089)

* Update add-approve-label.yml

* Update add-approve-label.yml

* Update add-approve-label.yml

caneGuy commented on be/src/formats/csv/array_reader.h:
```diff
@@ -0,0 +1,42 @@
+    [[nodiscard]] bool split_array_elements(Slice s, std::vector<Slice>* elements) const override;
+};
+
+// Reader for Hive text array format.
```
Please add some comments to explain what collection delimiter and mapkey delimiter do

ZiheLiu committed:
[Enhancement] Use object instead of pointer for HyperLogLog::_hash_set (#9073)

#6885 uses pointer instead of object for HyperLogLog::_hash_set to reduce memory usage of HLL and not include phmap.h in hll.h.

However, as for reading HLL column from disk,

using pointer will allocate memory 4096 times for HLL hash_set of 4096 rows in a chunk,
while using object only need allocate 1 times for a chunk by _pool.reserve.

zombee0 committed:
Remove unused InfoSchemaDb ctor (#9093)

zombee0 committed:
[Refactor] add default_cluster in LocalMetaData and Database so that can remove it from stmt safely (#9088)

add default_cluster in LocalMetaData and Database so that can remove it from stmt safely.

femiiii committed:
[Doc]Update sql statement (#9092)

wuleistarrocks committed:
Fill cache configuration before creating new cache filesystem (#9087)

* Fill cache configuration before creating new cache filesystem
if local disk cache enabled

* Disable local disk cache if cache_dir set to empty

HangyuanLiu committed:
Fix join precedence error (#9063)

The precedence of the comma operator is less than that of INNER JOIN, CROSS JOIN, LEFT JOIN, and so on
In consecutive joins without on conditions, a left-deep tree should be generated, and if there is an on predicate in the last join, it should correspond to the join formed by all preceding tables and the join generated by the last table.

Pslydhh committed:
[Enhancement] support BIGINT/INT argument for window_funnel (#9032)

Reduce copy when access array daum.
Add BIGINT and INT as argument of window_funnel to remove cast costs.

Smith-Cruise committed:
Refactor array_converter

Smith-Cruise committed:
Refactor array_converter

Seaven committed:
[Enhancement] CTE prune shuffle column (#8948)

murphyatwork committed:
Fix the thread-safety of phmap (#9099)

- `parallel_flat_hash_map::lazy_emplace` returns a iterator, which is not safe to read cause following insert would invalid it
- `parallel_flat_hash_map::lazy_emplace_l` 
  - It is safe to read if exists, or insert a value if not exists
  - This function would exclusively lock the inner map during the lookup or insert

abc982627271 committed:
Fix truncate lake table bug (#9002)

After truncated , showing partition and tablet would return 'unknown error', this pr fix this bug

Youngwb committed:
[BugFix] Fix CompoundPredicateOperator not equals when same children with different order (#8810)

sduzh committed:
[Enhancement] lake tablet GC (#9074)

murphyatwork committed:
[Enhance] introduce unplug mechanism to improve scalability (#8979)

sevev committed:
[Enhancement] Reject create table request if checkPersistentIndex failed (#9021)

PersistentIndex is available in branch-2.3, but there are many limitations in key columns. For example, key column is not supported as varchar(char) type, and the length of key columns cannot exceed 64 bytes.

In the previous implementation, if we create a primary key table enable PersistentIndex but the restrictions are not met, we will use PrimaryIndex in memory but not PersistentIndex, and primary key table will be create successfully. This may be a bit confusing to the user.

So we will reject the create table request directly if we find that the restrictions are not met.

padmejin committed:
[Enhancement] Add UT tools to verify the procedures of replay & image (#8943)

1. Add PseudoImage to help to verify the procedure of image on metadata.
2. Add PseudoJournalReplayer to help verify the procedure of replay on metadata.
Use the above tools on some unit test as demo.

ABingHuang committed:
[BugFix] Change the refresh task type of load-triggered materialized view to EVENT_TRIGGERED (#9090)

Change the refresh task type of load-triggered materialized view to EVENT_TRIGGERED

padmejin committed:
[BugFix] use findValidPort() to fix unstable ut caused by port conflict (#9111)

decster committed:
[Enhancement] Add PseudoCluster (#8854)

This PR adds an PseudoCluster, it includes PseudoBackend and PseduoFrontend. PseudoBackend simulates basic BE functions like tablet/txn management, task execution, and reporting, etc. The purpose of this is to facilitate some FE function/feature testing, like load/txn, ddl & dml execution, clone/balance behavior etc.
Currently, it conflict with the existing UT code (mainly because their are too many singletons in FE code), so to use it, the user should invoke a separate new process:

```
mvn compile exec:java -Dexec.mainClass="com.starrocks.pseudocluster.PseudoCluster"

this will create a 3 node pseudo cluster, then connect to this cluster using MySQL client

mysql -h 127.0.0.1 -P 9030 -uroot
// create table & insert into
```

HangyuanLiu committed:
[BugFix] Fix bug min/max statistics query result error (#8779)

trueeyu committed:
[Enhancement] set enable.partition.eof default to true (#9054)

For transaction producer, producer will append one control msg to the group of msgs, but the control msg will not return to consumer, so we can't to judge whether the consumption has been completed by offset comparison.

So we set `enable.partition.eof=true`, if we find current partition is already reach end, we will direct return instead of wait consume until timeout.

Another advantage of doing this is that for topics with little data, there is no need to wait until the timeout and occupy a lot of consumer threads.

dirtysalt committed:
Use `append_continuous_strings` in FLBAPlainDecoder (#9110)

Strings in `FLBAPlainDecoder` are continuous.

```
    Status next_batch(size_t count, ColumnContentType content_type, vectorized::Column* dst) override {
        if (_offset + _type_length * count > _data.size) {
            return Status::InternalError(strings::Substitute(
                    "going to read out-of-bounds data, offset=$0,count=$1,size=$2", _offset, count, _data.size));
        }
        std::vector<Slice> slices;
        slices.resize(count);
        for (int i = 0; i < count; ++i) {
            slices[i] = Slice(_data.data + _offset, _type_length);
            _offset += _type_length;
        }

        dst->append_continuous_strings(slices);
        return Status::OK();
    }
```

starrocks-xupeng committed:
[Feature] support checkpoint with subdir (#8907)

support write checkpoint to subdir
support transfer subdir's image

decster committed:
[Bugfix] fix compile error] (#9122)

trueeyu committed:
[BugFix] Fix the bug of not set need_rollback to false (#9114)

the ctx->need_rollback = false; will not reach

imay commented on be/src/exec/vectorized/hdfs_scanner_text.cpp:
```diff
@@ -99,16 +99,26 @@ Status HdfsScannerCSVReader::_fill_buffer() {
     SCOPED_RAW_TIMER(&_stats.reader_init_ns);
-    for (int i = 0; i < _scanner_params.materialize_slots.size(); i++) {
-        auto slot = _scanner_params.materialize_slots[i];
+    for (const auto& slot : _scanner_params.materialize_slots) {
```
If slot is a raw pointer, needn't use a reference.
use `cont auto slot` directly.

nshangyiming committed:
[BugFix] retry on SHUTDOWN tablet occupied (#8873)

Add retry logic to handle clone failure caused by SHUTDOWN tablet occupied
Treat NOT_FOUND on dest replica as a normal case

sduzh committed:
Add drop table interface between FE and BE for lake table (#9117)

xlwh committed:
[Refactor] Remove unused parse and analyze code in old parser (#9125)

Smith-Cruise committed:
Update hdfs_scanner_text.cpp

Smith-Cruise committed:
Update hdfs_scanner_text.cpp

xlwh committed:
Fix compilation failure issue (#9127)

Seaven committed:
[Enhancement] Update some class path (#9100)

zombee0 committed:
[Refactor] Remove default_cluster from db related stmt (#9130)

kevincai committed:
Fix StarletLocationProvider ut failure (#9132)

Signed-off-by: Kevin Xiaohua Cai <caixiaohua@starrocks.com>

trueeyu committed:
Upgrade librdkafka from 1.7.0 to 1.8.2 (#9133)

On version 1.7.0, query_watermark call is not transaction-aware. If we use read-committed, we will get the MaxOffset not the Last Stable offset. FE recorded the Last Stable offset but got MaxOffset, FE try to consume Kafka because of the lag of offset. But the read-committed transaction level will not consume any data. FE will retrying again and again. This comes from the producer write large of message but not committed. libkafka 1.8.2 resolves this problem and we upgrade to it.

edenhill/librdkafka#3423

stdpain committed:
[Enhancement] optimize aggregate performance for nullable column (#9080)

avoid call virtual function in NullableAggregateFunctionUnary, then make
nested function inline some function such as update/finalize

meegoo committed:
[Feature] Support prepare transaction stream load interface for 2PC (#8403)

StarRocks implements the /api/transaction/prepare interface to put the current transaction to be PREPARED state which temporarily persists changes. After preparing a transaction, the user can continue to commit or rollback the transaction.

The transaction preparation interface helps users achieve a more robust cross-system two-phase commit. After the transaction preparation is successful, even if StarRocks crashes, the user can continue to perform the commit or rollback this transaction after the system is recovered.

Smith-Cruise committed:
Unify file suffix (#9138)

stephen-shelby committed:
[Feature] Support use catalog (#9069)

we want to support "use catalog <catalog_name>".
If you send query "use catalog <catalog_name>" from mysql client, it will be parsed by mysql client to "use catalog" due to COM_INIT_DB protocol. so you need to execute "use 'catalog <catalog_name>'".
The query from jdbc or bi tools will be parsed normally by COM_QUERY protocol.

padmejin committed:
[BugFix] When forwarding SQL to leader, forward all modified session variable as well (#8966)

1.Store modified session variables to ConnectContext.
2.Add a SetStmt SQL to LeaderOpRequest to set all modified session variables
3.Leader will apply all session variables before executing the forwarded SQL.

meegoo committed:
[Enhancement] Support broker load non nested array type from parquet file (#9131)

gengjun-git committed:
[BugFix] Fix the bug of lose catalog  (#8998)

We should make sure the catalog is copied to majority of fe (write log success) before the catalog can be seen.

gengjun-git committed:
[Refactor] [Step 2/2]Change the name from master to leader (#8954)

Change the master related code to leader.
Split the task into two steps: 
1. change the name from master to leader in the internal code paths.
2. change the name from master to leader in the interface.

The advantage of this is that we only need to rollback the step 2 if If there are compatibility problems.

This is PR is the step 2 of the task. 
1. remove the `isMaster` attribute from the result of `show frontends`.
2. add `forward_to_leader` session variable to replace the `forward_to_master`. But we still remain `forward_to_master` for compatibility.

kangkaisen committed:
[Refactor] Remove some old Analyzer codes finally (#9107)

padmejin committed:
[Bugfix] retry when fail to start a local bdb cluster (#9149)

UT on the BDB cluster will sometimes fail when there are a lot of job running and the load are very high and will succeed most of the time after retry. I'm adding an automatic retry when creating the cluster.

sevev committed:
[BugFix] Primary key length is inconsistent with be (#9148)

PersistentIndex is available in version 2.3, but there are many limitations in key columns. For example, key column is not supported as varchar(char) type, and the length of key columns cannot exceed 64 bytes. However, the calculate logic in fe is inconsistent with be right now(DATE/DATETIME), so we may reject some create request actually we can do.

wanghuan2054 committed:
Support anltr4 for ShowOpenTableStmt #7977 (#8753)

zombee0 committed:
[Enhancement] Introduced brpc-java as brpc client (#8914)

Intruduced brpc-java as brpc client, so that callback can be used.

murphyatwork committed:
Fix the deadlock of resource group (#9147)

Deadlock:
- When register metric for resource group: take the `WorkGroupManager::mutex`, then `MetricRegistry::mutex`
- When update metric for resource group: take the `MetricRegistry::mutex`, then `WorkGroupManager::mutex`

As a fix, release the `WorkGroupManager::mutex` when registering metric, and using the return value to indicate register success or not.

ABingHuang committed:
[Enhancement] Fix refresh materialized view when materialized view is not active (#9084)


make refresh materialized view fail when materialized view is not active

hellolilyliuyi committed:
[docs]modify navigation (#9168)

Smith-Cruise committed:
Refactor again

Smith-Cruise committed:
Refactor again

nshangyiming committed:
[BugFix] set enable_strict_storage_medium_check back to false (#9182)

hellolilyliuyi committed:
[Docs]modify navigation (#9177)

Astralidea committed:
[Refactor] Refactor materialized view partition synchronize logical (#9022)

Separate the logic, first calculate the partitions that need to be added and deleted, and then perform the corresponding actions.
Remove tableMvPartitionNameRefMap & mvTablePartitionNameRefMap use rule-based partition mapping.
Use the new algorithm to calculate the corresponding partition
Consider the issue of flush propagation

trueeyu committed:
[Enhancement] Commit offset if the kafka offset moved (#9121)

If the producer start one transaction, and then write two msg (offset 0, offset 1) to kafka but not commit, then sr consumer use read-uncommitted to read, it will consume two msg (offset 0, offset 1).

Then producer commit, it will write one control msg (offset 2) to kafka.

In the next schedule, sr consumer consume 0 message, it will cancel the transaction, so the fe will keep retrying.

So if will consume no messages but offset is changed, we should commit one empty transaction to let fe change it's record offset

zombee0 committed:
[Refactor] Remove default cluster from some stmt (#9188)

murphyatwork committed:
[BugFix] fix shared-scan check (#9044)

sduzh committed:
Remove load channel clean thread (#9140)

There are too many threads in our BE process, try to reduce the number of threads by
running some lightweight tasks in bthread instread of pthread.

stephen-shelby committed:
[Refactor] Adapt to refresh hive external table when processing some consistent cases. (#9144)

Add writelock to prevent dirty read when processing hive external table schema change.
Prevent data race when refreshing at the same time.
Prevent schema of table be cleared caused by type conversion failure

stdpain committed:
[Enhancement] Improving memory locality of traversal tables (#8806)


Now the order in which we traverse the hash table is directly to the hash table.
Although direct access to flat hash table is a fast operation. He has very good memory locality.
but in fact, the value of hash table is a pointer.
eg:
[(k1, s1), (k2, s2)]
the access order is (k1,s1), (k2, s2).
But s1 and s2 may be farther apart from each other.
Because the order of our hash insertion is not a fixed order.
To improve the speed of hash table traversal. We choose to save both keys of the hash table in the state (maximum sizeof(key)==16).
At the same time we save the creation order of the states (states are allocated consecutively from the pool, the same creation order has better memory locality).
When outputting the hash table, we output the hash table by the traversal order of the states

performance case:
baseline 2.3-rc02
```
1FE 1BE pipeline items:99997497
SELECT SUM(IsRefresh), COUNT(*) AS c, AVG(ResolutionWidth) FROM hits GROUP BY WatchID, ClientIP ORDER BY c DESC LIMIT 10;
6.58 -> 4.85
SELECT URL, COUNT(*) AS c FROM hits GROUP BY URL ORDER BY c DESC LIMIT 10;
6.53 -> 5.44
```

femiiii committed:
Update External_table.md (#9195)

stdpain committed:
[Bugfix] fix window_funnel crash when input array has null (#9191)

if event_list was empty, serialize should append a empty array

decster committed:
[Bugfix] clear rowsets before load (#9193)

SchemaChange will call _load_from_pb to reload metadata, but _load_from_pb forgets to reset _rowsets vector, so some old rowsets may still be stored in _rowsets, causing no delvec found error, because the delvec corresponds to the old rowset is already removed. This PR clears _rowsets vector before load new rowsets from meta.

padmejin committed:
[BugFix] allow deprecated variables forward to master by default (#9196)

Fix NPE when trying to get flag from deprecated variables such as 'enable_vectorized_engine' and 'enable_cbo'

Smith-Cruise committed:
Rewrite array_converter.cpp.

Signed-off-by: Smith Cruise <chendingchao1@126.com>

Smith-Cruise committed:
Rewrite array_converter.cpp.

Signed-off-by: Smith Cruise <chendingchao1@126.com>

Smith-Cruise committed:
Merge branch 'fix' of github.com:Smith-Cruise/starrocks into fix
